pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -541,17 +554,52 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 		return err
 	}
 
-	// Anything left in addrs is a new member that needs to be added to a pool
-	for addr := range addrs {
-		_, err := v2_pools.CreateAssociateMember(lbaas.network, poolID, v2_pools.MemberCreateOpts{
-			Address:      addr,
-			ProtocolPort: int(ports[0].NodePort),
-			SubnetID:     lbaas.opts.SubnetId,
-		}).Extract()
-		if err != nil {
-			return err
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+	for _, port := range ports {
+		var listenerID, poolID string
+
+		// Find appropriate listener for this port
+		for _, listener := range listenersList {
+			if listener.ProtocolPort == int(port.Port) && listener.Protocol == string(port.Protocol) {
+				listenerID = listener.ID
+				break
+			}
+		}
+		if listenerID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required listener for port %d and protocol %s", loadBalancerName, port.Port, port.Protocol)
+		}
+
+		// Find appropriate pool for found listener
+		for _, pool := range poolsList {
+			for _, currentListener := range pool.Listeners {
+				if listenerID == currentListener.ID {
+					poolID = pool.ID
+					break
+				}
+			}
+			if poolID != "" {
+				break
+			}
+		}
+		if poolID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required pool for listener %s", loadBalancerName, listenerID)
+		}
+
+		for addr := range addrs {
```

Comment:
the update has been cleaned up.  I still have the remove non-required members and add new members as separate pieces.  Reason being when removing members, we don't care which port that member was for, therefore there is no reason to loop through ports then.  When adding new members, since new members are added in respective pools with port information, we need to iterate over the set of ports this lb is serving.